### ISSUE TITLE
LG-313: MicroProfile Health 3.1 Implementation

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#16626]
## Description of the high level feature, including any external spec links:  
Implement the MicroProfile Health 3.1 API/Spec/TCK in Open Liberty

##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/lbls0vuyozr1xzr8o4oht8kq2jh32isx
- FTS: https://github.com/OpenLiberty/open-liberty/issues/17550
- Beta Blog Post (if applicable): https://openliberty.io/blog/2021/07/12/logging-format-21008-beta.html
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/17942
- ID: https://github.com/OpenLiberty/docs/issues/4327

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [x] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [x] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by frowe** [comment]
MicroProfile Health 3.1 UFO review notes 5/26/2021

Add /health/start endpoint to 2nd bullet of slide 27 Security
Remove reference to specific beta date,21.0.0.8, on slide 22

**Comment by pgunapal** [comment]
UFO has been updated with the review notes from today's (26/05/2021) UFO socialization meeting.

**Comment by pgunapal** [comment]
Issue created for OL guides updates: https://github.com/OpenLiberty/guides-common/issues/683

**Comment by NottyCode** [comment]
@pgunapal the UFO doesn't say what happens if there is not `@Startup` health check, and the application isn't started. I know it kind of seems obvious, but this is something that has been got wrong in the past so I'd like it to be called out explicitly in the UFO that if the application is not yet started then the `/health/started` endpoint will be DOWN.

**Comment by pgunapal** [comment]
@NottyCode I have updated Slide 13 in the UFO to explicitly call out what happens if there are no `@Startup` health checks and the application isn't started yet. The `/health/started` endpoint will be DOWN with an empty checks response, until the application has started.

**Comment by pgunapal** [comment]
@samwatibm All the translated messages are in the release branch, may I get the Globalization approval for this feature?

**Comment by pgunapal** [comment]
@steven1046 Hi Steven, There are no addition of any new or change to any existing UI. May I get your approval on Accessibility? Thanks!

**Comment by pgunapal** [comment]
@skasund Hi Swati, I have completed and uploaded the STE for mpHealth-3.1 (https://ibm.box.com/s/5wju64id742f1wn2l76stni0rq892oby). May I get your approval for STE?

**Comment by pgunapal** [comment]
@tevans78 Hi Tom, the demo is scheduled for the upcoming EOI meeting (July 20th). May I get the Demo approval for this feature?

**Comment by chirp1** [comment]
Prashanth's info added to the ID issues at https://github.com/OpenLiberty/docs/issues/4327 and the WL ID issue look good. Approving this epic.

**Comment by skasund** [comment]
@pgunapal   Thanks for the STE slides.

**Comment by pgunapal** [comment]
@donbourne **Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

- Yes, the UFO includes the most likely problems the customers will see, which is usually in their own implementation of the HealthCheck interface, in their applications. Detailed messages will be logged in the messages.log. Major functional changes in mpHealth-3.1 were the addition of the new `@Startup` annotation and the new REST endpoint `/health/started`, its implementation is similar to the existing annotations (Liveness/Readiness) and REST endpoints in MP Health.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated?
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

- For mpHealth-3.1, the new `@Startup` annotation was tested if users did not specify that annotation in their HealthCheck implementations, the appropriate error message will be shown. The new REST endpoint `/health/started` would show the appropriate HTTP response code status and message, if the endpoint can not be reached, or if the Health Checks failed or if the deployed applications have not started yet. These problematic scenarios already existed for the existing annotations and endpoints., and was tested and demo'ed. This feature will be demo'ed in the upcoming EOI meeting.

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?

-  Daniel Guinan from Brian Hanczaryk's team had done SVT testing for this feature, where they had updated their applications to use mpHealth-3.1 feature, along with the `@Startup` annotation and there weren't any issues. They had also tested configuring the new `/health/started` endpoint to the Kubernetes Startup Probe, in OCP, and that successfully passed as well.

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

- Yes.

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
- WAS L3: Liberty Log Analytics
